### ISSUE TITLE
Allow key-spacing align to align to nearest indentation level instead of a single space

### ISSUE THREAD

**Issue by daedalus28** [eslint/eslint#3453]
The key-spacing align works great for this case:

    var x = {
        a: 1,
        b: 2
    }

But in my code, I prefer to align to the nearest indentation level (which in this case is 4-spaces). Note that there are two spaces between the colon and the value - this is because it would take two spaces to get it to the next indentation level (multiple of 4 spaces)

    var x = {
        a:  1,
        b:  2
    }

It's easier to just hit tab until things align instead of some combination of tab and spaces (in editors that support soft tabs), and it also leaves more room for objects to grow without having to modify every key value pair.

The simplest way would probably be passing in the tab size to this rule, but it would be neat if it could read the value from the indent rule.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by daedalus28** [comment]
I also want to point out that the second case also causes key-spacing to throw an error complaining that there are too many spaces. If nothing else, a simple flag to ignore extra whitespace provided all values are aligned would at least prevent eslint from throwing errors (preventing me from enforcing this rule)

**Comment by nzakas** [comment]
@btmills thoughts?

**Comment by BYK** [comment]
This seems a bit related to #3363

**Comment by btmills** [comment]
If we had an option to the effect of "I want my (values|colons) aligned, but I don't care where", would that adequately cover this and #3363?

**Comment by btmills** [comment]
To be more specific: If we had an option, "spacing", which accepted values "exact" (default, current behavior) or "any" (true means "at least one"), would that cover this?

**Comment by daedalus28** [comment]
The spacing option would be a great start and make the rule usable, but what I'd really like is to force alignment to the lowest possible indentation level that preserves alignment. The spacing option will stop it from throwing errors for our intended use case, but won't check to see if developers are aligning halfway between tab stops.

I think a great solution would be to add a third possible value for the spacing specifying indentation level like the indent rule currently works - a number representing the number of spaces or the string  "tab" specifying tabs are used. Even better would just be specify "indent" and have it look it up from the indent rule's config (if that's possible)

**Comment by BYK** [comment]
I like the `spacing` option idea @btmills! We can actually expand the existing `beforeColon` and `afterColon` settings to be numbers after that.

@nzakas - what do you think?

**Comment by nzakas** [comment]
I think it's the right direction, though I think `spacing` is an ambiguous term, so I'd like to think about something else.

**Comment by BYK** [comment]
How about `colonPadding`?

**Comment by nzakas** [comment]
That sounds like it applies to both before and after the colon, but I think we're looking at only caring about after here?

**Comment by btmills** [comment]
I think it depends whether `align` is set to `value` or `colon`.

**Comment by nzakas** [comment]
@btmills can you explain more? Do you have another suggestion for the option name?

**Comment by btmills** [comment]
Assuming four-space indentation, configured via a TBD option not included in the examples:

```js
/* eslint key-spacing: [2, { "align": "value" }] */
({
    abcd:   "hello",
    ef:     "world"
});
```

Here, the combination of `beforeColon`'s default `false` value and `value` alignment means that there should never be space before the colon - all padding and alignment occurs between the colon and the value.

```js
/* eslint key-spacing: [2, { "align": "colon" }] */
({
    abcd: "hello",
    ef  : "world"
});
```

`beforeColon` still defaults to `false`, but `colon` alignment means there will be whitespace for alignment between the key and the colon.

Where we add the alignment whitespace depends on the value of `align`:

- `value`: between the colon and the value
- `colon`: between the key and the colon

@daedalus28 looking at my second example, I aligned the colons at four spaces. In the case of `colon` alignment, would you want the colons or the values to be at the next indent level?

**Comment by btmills** [comment]
> Do you have another suggestion for the option name?

Perhaps `mode` and a string value of `strict` or `minimum`?

**Comment by nzakas** [comment]
I like it!

**Comment by daedalus28** [comment]
@btmills I would think that in colon alignment, the colons would be at the next indentation level. The point is the thing that's aligned is aligned at the next indentation level instead of with arbitrary numbers of spaces

**Comment by IanVS** [comment]
I like the direction this is going, but would not want to force indentation level, which I think is overly complicated.  I myself like to use colon alignment, but with a space between the key and colon (which currently can't be done).  If there is a `mode` which can allow arbitrary spacing, that would be sufficient for me.  Otherwise, I fear the option might become confusing.

**Comment by btmills** [comment]
> with a space between the key and colon (which currently can't be done)

Why doesn't `"beforeColon": true` work? Other than that, I'm in agreement.

**Comment by IanVS** [comment]
Ah, you're right.  I learn something new every day.

If we're allowing arbitrary spacing, will it have to be 1+ spaces, or 0+?  If 0+, it would seem to address https://github.com/eslint/eslint/issues/3120.

**Comment by btmills** [comment]
1+, since we need a way to prevent spacing still. I have yet to see a good way to solve #3120 with the key-spacing rule that exists in core.

**Comment by alberto** [comment]
Closed by lack of activity
